core/parser.mly

```diff
@@ -169,7 +169,10 @@ let parseRegexFlags f =
               | 'g' -> RegexGlobal
               | _ -> assert false) (asList f 0 [])
 
-
+let switch_fun_currying_guard p args =
```

Comment:
It is better to define this in the desugar pass module such that you only need to call it once.

Reply:
Since I have changed the type of `switch_funlit` to `Pattern.with_pos list * switch_funlit_body`, I think it is a syntax error and should be checked in parsing phase?

Reply:
Conceptually it is a syntax error, and thus I understand why you would wire this check into the parser. Syntax errors do not exclusively arise during parsing as sometimes you can choose to relax the implemented language grammar a bit in favour of, say, a simpler parser. To catch the false-positives produced by the parser you can subsequently traverse syntax tree to check that it has the intended form.

The duplication of function definition production rules in the parser means you have multiple call sites for `switch_fun_currying_guard`. You can get away with having only one call site if invoke along with the other guards in the desugaring pass. This design is leaner,  and will make the code base more robust for future changes, say, if we need to duplicate rules in the parser again.

Reply:
I see. Consider we may support currying for it in future. I think its better to change the type of `switch_funlit` back to `Pattern.with_pos list list * switch_funlit_body` and put the guard in `desugarSwitchFuns` pass to raise an error if the user gives it multiple parameter list. (If that's a bad design please correct me)

Reply:
I think it is fine to slightly generalise the shape `switch_funlit`.